File: 2023/04.25/서지훈_백준_11725_트리의_부모_찾기.py
import sys
sys.setrecursionlimit(1000000)


def dfs(node):
    for next_node in adjacent[node]:
        if not visited[next_node]:
            visited[node] = True
            parent[next_node] = node
            dfs(next_node)


N = int(input())
adjacent = [[] for i in range(N+1)]
for i in range(N-1):
    a, b = map(int, input().split())
    adjacent[a].append(b)
    adjacent[b].append(a)

parent = [0] * (N+1)
visited = [False] * (N+1)
visited[1] = True
dfs(1)

for i in range(2, N+1):
    sys.stdout.write(str(parent[i]) + "\n")

File: 2023/04.25/서지훈_백준_11725_트리의_부모_찾기.cpp
#include <iostream>
#include <vector>

int parent[100001];
std::vector<int> adjacent[100001];
bool visited[100001];

void dfs(int node){
	for(auto next_node : adjacent[node]){
		if(!visited[next_node]){
			visited[next_node] = true;
			parent[next_node] = node;
			dfs(next_node);
		}
	}
}

int main(){
	int N;
	std::cin >> N;
	
	for(int i = 0; i < N-1; ++i){
		int a, b;
		std::cin >> a >> b;
		adjacent[a].push_back(b);
		adjacent[b].push_back(a);
	}
	
	visited[1] = true;
	
	dfs(1);
	
	for(int i = 2; i <= N; ++i){
		std::cout << parent[i] << '\n';
	}
}